### ISSUE TITLE
F# support is broken in Xamarin.Android

### ISSUE THREAD

**Issue by Dolfik1** [dotnet/fsharp#12640]
**Repro steps**

Provide the steps required to reproduce the problem:

1. Install Visual Studio or Visual Studio for Mac and enable both Xamarin and .NET Core support, these are listed as ‘Mobile development with .NET’ and ‘.NET Core Cross-platform development’ respectively.

2. Open a command prompt window and install the template pack by entering:
```dotnet new -i Fabulous.XamarinForms.Templates```

3. Navigate to a folder in the command prompt window where your new app can be created and enter:
```dotnet new fabulous-xf-app -n SqueakyApp```

4. Build and run SqueakyApp.Android on device.

**Expected behavior**

The app build and run successfully

**Actual behavior**

```
/Users/nikolay/Dev/fsharp/SqueakyApp2/SqueakyApp2.Android/AssemblyInfo.fs(9,30): error FS0039: The type 'Android' is not defined.
/Users/nikolay/Dev/fsharp/SqueakyApp2/SqueakyApp2.Android/AssemblyInfo.fs(9,30): error FS0039: The type 'Android' is not defined.
/Users/nikolay/Dev/fsharp/SqueakyApp2/SqueakyApp2.Android/MainActivity.fs(20,63): error FS0039: The value, constructor, namespace or type 'Layout' is not defined.
/Users/nikolay/Dev/fsharp/SqueakyApp2/SqueakyApp2.Android/MainActivity.fs(21,61): error FS0039: The value, constructor, namespace or type 'Layout' is not defined.
```

These errors are due to Xamarin.Android.FSharp.ResourceProvider does not compile and does not load Resources library:
```type Resources = SqueakyApp2.Android.Resource```

**Known workarounds**

Use Paket instead of NuGet. Looks like Paket in a magical way works with this. 
But Paket does not support MonoAndroid11 and MonoAndroid12 ([link1](https://github.com/fsprojects/Paket/issues/4112), [link2](https://github.com/fsprojects/Paket/pull/4113))

I also found out that there is no way to add back Xamarin.Android.FSharp.ResourceProvider 1.0.0.28 after remove:
```
Could not install package 'Xamarin.Android.FSharp.ResourceProvider 1.0.0.28'. You are trying to install this package into a project that targets 'MonoAndroid,Version=v10.0', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.
```

I tried to install `Xamarin.Android.FSharp.ResourceProvider 1.0.1` and it is installed successfully but this version does not work ([link1](https://github.com/xamarin/Xamarin.Android.FSharp.ResourceProvider/issues/11), [link2](https://github.com/fsprojects/FSharp.TypeProviders.SDK/issues/376)) and I getting the same error as above with warning:
```
 Referenced assembly '/Users/nikolay/.nuget/packages/xamarin.android.fsharp.resourceprovider/1.0.1/lib/monoandroid81/Xamarin.Android.FSharp.ResourceProvider.Runtime.dll' has assembly level attribute 'Microsoft.FSharp.Core.CompilerServices.TypeProviderAssemblyAttribute' but no public type provider classes were found
```

**Related information**

Provide any related information (optional):

* macOS 12.1
* Xamarin.Android Version: 12.0.0.3 (Visual Studio Community)
* Visual Studio Community 2019 for Mac Version 8.10.16 (build 2)

cc @dsyme @vzarytovskii @nosami

**Comment by Happypig375** [comment]
Duplicate of https://github.com/dotnet/fsharp/issues/10837, https://github.com/xamarin/Xamarin.Android.FSharp.ResourceProvider/issues/9, and https://github.com/xamarin/xamarin-android/issues/6404

**Comment by TimLariviere** [comment]
In general, the state of mobile development with F# is a real mess right now.

Just to name a few:
- All the issues with `Xamarin.Android.FSharp.ResourceProvider` as mentioned here (`1.0.1` never worked)
- Existing F# Xamarin templates removed from Visual Studio 2022
- Mono not supporting F# 6.0
- No official way to use F# with `net6.0-ios` and `net6.0-android` (I don't even know if we can create Android project with F# on .NET 6.0)

Sorry for hijacking this issue and venting my frustration, but we could really use some help from Microsoft on the F# + Mobile front. :)

**Comment by Dolfik1** [comment]
>     * No official way to use F# with `net6.0-ios` and `net6.0-android` (I don't even know if we can create Android project with F# on .NET 6.0)

I have successfully ported our F# mobile app to net6.0-ios. It seems most of code works without any issues. I also successfully ran sample app on net6.0-android (but without using android resources).

**Comment by 7sharp9** [comment]
@TimLariviere These things tend to happen people move to different companies and depertments and were never replaced, things enter maintenance mode and are forgotten etc.  A lot of the F# integration used to be open source, but Im not sure if parts were made internal after the acquisition.  I think the F# droid stuff was internal to be honest now I think back to when I was there.

**Comment by 7sharp9** [comment]
The type provider could be replaces by Myriad too to make it easier to maintain in the future and open source.

**Comment by nosami** [comment]
Xamarin.Android.FSharp.ResourceProvider was originally written by myself because of issues with the F# and CodeDOM provider that came before it that couldn't easily be fixed (F# doesn't support public static fields and Xamarin.Android required these). The problem was that I was never supposed to work on that as I was always part of the VSMac team not Xamarin.Android.

I later moved to the Codespaces team which meant that I was even further removed from Xamarin.Android work. At the time that I moved, I handed over Xamarin.Android.FSharp.ResourceProvider for maintenance to the XA team but it seems like bitrot set in pretty quickly.

I've since returned to the VSMac team, but now that VSMac shares much of the same codebase as Visual Studio, there is no longer any need for dedicated F# staff on the Mac side of things so I mostly work on other areas such as the debugger but have spent a lot of time recently porting parts of the UI to Cocoa.

I could work on the type provider in my spare time, but I have very little of it. It is open source though so anyone could work on it if they wanted to.

I enquired about the state of play with MAUI and net6.0-android. The plan is to ditch the C# code generation completely (which the F# type provider uses in the background) and use IL generation instead which at least in theory should work for all dotnet languages. See [comment](https://github.com/fsprojects/FSharp.TypeProviders.SDK/issues/376#issuecomment-973923466) by @dellis1972  This should make the type provider completely redundant.

**Comment by TimLariviere** [comment]
Thanks for the replies

>I could work on the type provider in my spare time, but I have very little of it

I was not asking for you or anyone to work on their spare time, especially if it's not even your day job.

The message I wanted to convey was more like it would be really appreciated if Microsoft as a whole would consider .NET to not be only C#, and put on the roadmap some F# to get some basic support on everything labeled `.NET`.

Such an example is `PackageReference` in Visual Studio (Win) in old style fsproj.
VS Mac and Jetbrains Rider support it since a long time ago. VS Win doesn't which is a deal-breaker because most developers use this one.

More than 3 years ago, I tried to submit a PR to fix that but most of what was required was not open-source and the response I got was "Won't do it. Wait for Xamarin to move to sdk-style".
Now we're 2022, C# seems to have got it. F#, still very bumpy.

Sure, we could stay with `packages.config` and wait.
But since C# has moved away from that years ago, and all IDE tooling and Xamarin teams are slowly removing support for it.
"Just use PackageReference." Well, we can't :(

It's that kind of experience that makes it so frustrating.

> It is open source though so anyone could work on it if they wanted to.

I'll try to fix my main issues. Just hoping getting reviewed, merged and released won't take months since ownership is from teams at Microsoft...

@nosami Could I ping you if I happen to be stuck on non open-source stuff, so you could try to relay the message internally to the right people?

>I have successfully ported our F# mobile app to net6.0-ios. It seems most of code works without any issues. I also successfully ran sample app on net6.0-android (but without using android resources).

@Dolfik1 Would you happen to have an empty `net6.0-android` with F#?
I know there is this one for iOS: https://github.com/edgarfgp/Net6iOSTemplate
It would be great to release those templates on NuGet for both iOS and Android (and Xamarin.Forms too).

I tried on my own to convert from C# to F#, but got inflate errors...

**Comment by Dolfik1** [comment]
> F# doesn't support public static fields and Xamarin.Android required these

The generated resource file also contains `partial` class which is also not supported in F#



> I was not asking for you or anyone to work on their spare time, especially if it's not even your day job.

We can try to fix type provider on our own. I think the main problem is the wrong type provider's project configuration.

> The message I wanted to convey was more like it would be really appreciated if Microsoft as a whole would consider .NET to not be only C#, and put on the roadmap some F# to get some basic support on everything labeled .NET.

I am agree with @TimLariviere. F# reduces development time and provides the ability to write more compact and reusable code. But at the moment we have to spend a lot of time solving tooling/mono/AOT problems instead of solving business problems. At the moment we want to migrate to .Net 6 because some of our problems seem to be solved.

> @Dolfik1 Would you happen to have an empty net6.0-android with F#?

I will attach sample app but it is incomplete without type provider.

> I tried on my own to convert from C# to F#, but got inflate errors...

I also wrote a migration script from an old project. This script is not perfect but mostly works with few manual changes. (this script does not support packages.config, only Paket is supported)


[migrator.fsx.zip](https://github.com/dotnet/fsharp/files/7940121/migrator.fsx.zip)

[Net6Sample.zip](https://github.com/dotnet/fsharp/files/7940123/Net6Sample.zip)

**Comment by TimLariviere** [comment]
Thanks!
I have created a repository to centralize all F# 6.0 templates for mobile development:
https://github.com/fabulousfx/net6.0-mobile-fsharp

This will be helpful to package them for `dotnet new` and put them on NuGet.
Later, I'll include Xamarin.Forms and maybe MAUI if I get it to work.

**Comment by nosami** [comment]
> @nosami Could I ping you if I happen to be stuck on non open-source stuff, so you could try to relay the message internally to the right people?

Sure! You can email me at `jasonimison [AT] gmail [DOT] com` or `jaimison [AT] MS` or I have an email rule to notify me if I am mentioned on GitHub.

I think personally I would stay with F# 5 and Mono until MAUI comes out. This issue should be resolved automatically.

**Comment by gsomix** [comment]
> This will be helpful to package them for dotnet new and put them on NuGet.

@TimLariviere Hey, could you create issue in your repository? I'll take a look.

**Comment by charlesroddie** [comment]
"F# support is broken in Xamarin.Android" suggests that creating Xamarin.Android apps using F# has problems. The issue is much smaller, that creating a Xamarin.Android platform project in F# has problems.

Simplest solution is for ```dotnet new -i Fabulous.XamarinForms.Templates``` to generate C# Xamarin.Android projects which depend on an F# Fabulous Xamarin.Forms netstandard project.

Platform head projects are very tooling-dependent and F# projects are always going to be in a cycle of breaking and being fixed and lagging or non-existent tooling.

**Comment by Dolfik1** [comment]
> I think personally I would stay with F# 5 and Mono until MAUI comes out. This issue should be resolved automatically.

We used FSharp.Core 4.7.2 for a long time due to AOT issues on iOS. Last month we updated FSharp.Core to version 6.0.2 and it seems that most of the issues have been fixed. But we got new issues with VTable setup on Android in System.Text.Json. It looks like it has been fixed somehow in .Net 6.

Moreover, I do not really believe that all problems will be fixed on MAUI release.

**Comment by dsyme** [comment]
> The issue is much smaller, that creating a Xamarin.Android platform project in F# has problems..... Platform head projects are very tooling-dependent and F# projects are always going to be in a cycle of breaking and being fixed and lagging or non-existent tooling. (Unless sdk-style projects fix all problems?)

I actually agree with this. It has always been a losing fight to attempt to have the more exotic platform heads in F#.  The iOS and Android platform heads took forever to move to .NET SDK, for example (have they moved now?) which made it impossible to maintain them.

It's up to the F# community, but there is wisdom in following @charlesroddie's advice here and consider the F# platform head projects for Android and iOS deprecated in favour of focusing on the incorporation of F# libraries and component code into C#-project-headed Android and iOS apps (that is, solutions headed by relatively minimal C# projects for packaging and delivery).  There are almost certainly more important things to advocate, and it's a realistic, stable position.

**Comment by TimLariviere** [comment]
Even though I understand the reasoning behind this, I feel it's unfair to abandon F# heads because Microsoft will always only focus on C#.

We're actually extremely close to have something fully working.

iOS head already work,
Android head's only issue is `Resource.designer.cs` only generated for C# - once compiled for F# all is good.

Other issues are not related to the F# heads, but to F# as a whole.

And I fear this "give-up" will only lead in the future to an abandonment of F# support à la .NET Native; blocking all apps with F# libs from being published to the Windows Store because of tail-recursion. ("Too hard to implement and nobody uses F#, so we don't care")

>For example, the iOS and Android platform heads took forever to move to .NET SDK, for example (have they moved now?)

Yes, with .NET 6.0.
Even though Microsoft only provided C# heads with .NET SDK, F# heads also work out of the box.

**Comment by dellis1972** [comment]
> iOS head already work, Android head's only issue is `Resource.designer.cs` only generated for C# - once compiled for F# all is good.

The Android head `Resource.desinger.cs` issue will hopefully be solved when this PR https://github.com/xamarin/xamarin-android/pull/6427 is completed. It will remove the need for the Type Provider in favour of generating a Resource Assembly via IL directy. I have tested this locally with an F# Head project as it works as expected. 
Note we plan to enable this new system for .net 6 onwards.

**Comment by TimLariviere** [comment]
@dellis1972 That's a wonderful news. Would you have an estimation when this would be available?
Same as MAUI release (mid 2022) ?

**Comment by Dolfik1** [comment]
> > The issue is much smaller, that creating a Xamarin.Android platform project in F# has problems..... Platform head projects are very tooling-dependent and F# projects are always going to be in a cycle of breaking and being fixed and lagging or non-existent tooling. (Unless sdk-style projects fix all problems?)
> 
> I actually agree with this. It has always been a losing fight to attempt to have the more exotic platform headed by F# projects. For example, the iOS and Android platform heads took forever to move to .NET SDK, for example (have they moved now?) which made it impossible to maintain them.
> 
> It's up to the F# community, but there is wisdom in following @charlesroddie's advice here and consider the F# platform head projects for Android and iOS deprecated in favour of focusing on the incorporation of F# libraries and component code into C#-project-headed Android and iOS apps (that is, solutions headed by relatively minimal C# projects for packaging and delivery). There are almost certainly more important things to advocate, and it's a realistic, stable position.

I am really sorry to hear that. We are developing a mobile application on F# for a few years. During this time we have contributed to F# on our own (Fabulous, Paket, Fantomas, FSharp.Core, LiteDb.FSharp, XF, Xamarin.Android, Xamarin.iOS etc). We are actively involved in the Russian F# community and hiring engage people. We made a Fabulous fork and develop it separateley but sharing experience with @TimLariviere to include this in the main repository. We also developed Fabulous.Android/Fabulous.iOS. This allows you to build Android/iOS apps with native UI on F# and it works pretty fine. We also have plans to make them open source. During this time we have included many people in the development on F# (we have retrained several people from Kotlin/Swift/Obj-C/C# to F#).

Our application, in addition to business logic, also includes a very complex UI. It's the reason why we do not want to C#-project-headed Android and iOS apps. For the same reason, we refused Xamarin.Forms. It is very painful to write a complex UI with Xamarin.Forms.

If you are interested, I can give you more details about our app and our architecture in PM.

**Comment by dsyme** [comment]
@dellis1972 That's great news

@Dolfik1 This is very valuable feedback - it's really important to understand the breadth of usage and all sceanrios being impacted

> Our application, in addition to business logic, also includes a very complex UI. It's the reason why we do not want to C#-project-headed Android and iOS apps. For the same reason, we refused Xamarin.Forms. It is very painful to write a complex UI with Xamarin.Forms.

Just to check, you're saying it's not possible to write this as a set of F# Android/iOS libraries implementing the complex UI, with a very thin C# head for resources, packaging etc? 

I'm actually not taking any specific opinion here, and will leave the quetion of head-support to various teams who have that responsibility - just agreeing with @charlesroddie's general point that historically they've been difficult.

**Comment by dsyme** [comment]
Well, it would be good to get issues raised in the correct repos - literally none of the above is implemented in this repo, so this is just not the right place to be discussing this.  The product owners for Xamarin.Android etc. are ultimately responsible here and you need to be talking with them.

**Comment by dsyme** [comment]
> Mono not supporting F# 6.0

@nosami Do we have a canonical issue link for this?

> No official way to use F# with net6.0-ios and net6.0-android (I don't even know if we can create Android project with F# on .NET 6.0)

Anyone have a link to any specific outstanding issues on this?

> Existing F# Xamarin templates removed from Visual Studio 2022

Anyone got any links to this?  TBH I'm not too surprised by this given the old templates used the old project format, and were VS-only, and I think actually only VS-for-Windows (or was it VS-for-mac?).   They'd need to be replaced by command line `dotnet new` templates.

**Comment by nosami** [comment]
>Mono not supporting F# 6.0

> > @nosami Do we have a canonical issue link for this?

Not as far as I know. Mono is being superceded by .Net Core.

**Comment by dsyme** [comment]
> Not as far as I know. Mono is being superceded by .Net Core. There are no plans to support F# 6 in Mono.

Right, I think that's part of the issue here - it's been hard to justify the udpate when Mono is being replaced within Xamarin stack.  So much chicken and egg.

> I'm currently getting F# 6 working in VS Mac, but that's a separate issue.

❤️ ❤️ ❤️ ❤️

**Comment by Dolfik1** [comment]
> Well, it would be good to get issues raised in the correct repos - literally none of the above is implemented in this repo, so this is just not the right place to be discussing this. The product owners for Xamarin.Android etc. are ultimately responsible here and you need to be talking with them.

The main problem at this moment it is the type provider. I have created an [issue](https://github.com/xamarin/Xamarin.Android.FSharp.ResourceProvider/issues/11) at November 12 but still no solution. For this reason, I decided to create an issue in this repository to draw more attention to the problem.

> Anyone got any links to this? TBH I'm not too surprised by this given the old templates used the old project format, and were VS-only, and I think actually only VS-for-Windows (or was it VS-for-mac?). They'd need to be replaced by command line dotnet new templates.

This is not possible at the moment (at least for android) due to type provider issues. So we first need to solve type provider problem. I really like what @dellis1972 does but I am afraid at the moment this cannot be a solution, since it is not known how soon it will be released. For this reason, I think we still need to solve the problem with the type provider.

However at the moment, nothing prevents you from create templates for iOS.

**Comment by dsyme** [comment]
> The main problem at this moment it is the type provider. I have created an issue at November 12 but still no solution. For this reason, I decided to create an issue in this repository to draw more attention to the problem.

What if you just don't use the type provider and address the resources by string? i.e. the generated code that the type provider produces?

**Comment by Dolfik1** [comment]
> What if you just don't use the type provider and address the resources by string? i.e. the generated code that the type provider produces?
It seems that this is really possible, but it is very expensive. It will look something like this:

```context.Resources.GetIdentifier("Widget_MaterialComponents_TextInputLayout_OutlinedBox", "style", context.PackageName)```

instead of

```
Resources.Style.Widget_MaterialComponents_TextInputLayout_OutlinedBox
```

This way is also unsafe because you can make a typo in resource name. It is also more difficult to refactor.

**Comment by dsyme** [comment]
> This way is also unsafe because you can make a typo in resource name. It is also more difficult to refactor.

Yes, but it works and unblocks.  

TBH I was always skeptical about putting a type provider into the main path of such a corner case of the Xamarin build story. I would have absolutely no problem with ripping it out of any defaults and making it a community-provided optional add-on that the community maintains - we don't bake in type providers like FSharp.Configuration into .NET Core projects even though they are also useful.   If we expect the Xamarin team to maintain F# support we need things to be really simple for them.

**Comment by Dolfik1** [comment]
> TBH I was always skeptical about putting a type provider into the main path of such a corner case of the Xamarin build story. I would have absolutely no problem with ripping it out of any defaults and making it a community-provided optional add-on that the community maintains - we don't bake in type providers like FSharp.Configuration into .NET Core projects even though they are also useful. If we expect the Xamarin team to maintain F# support we need things to be really simple for them.

You are absolutely right. But for now, we have what we have. There is good news about IL generated assembly for resources but we need but still we need some temporary solution. 

At the moment @TimLariviere is trying to revive the type provider but there are some [difficulties](https://github.com/fabulousfx/FSharp.Android.ResourceProvider/issues/1).

**Comment by TimLariviere** [comment]
FYI: I abandoned trying to revive the type provider (too many issues with it) and did a small MSBuild target to build the C# designer as part of the build.

https://github.com/fabulousfx/FSharp.Android.Resource

A little rough for the moment, but it works with both .NET Framework and .NET 6.0!

Should be good enough until we get https://github.com/xamarin/xamarin-android/pull/6427

**Comment by nbevans** [comment]
That's a very simple solution @TimLariviere one wonders why the original impl didn't just do this?  Looking forward to the IL generation though.

**Comment by dsyme** [comment]
> That's a very simple solution @TimLariviere one wonders why the original impl didn't just do this? Looking forward to the IL generation though.

It should have, a type provider wasn't the right solution for this case, especially in retrospect.

**Comment by dsyme** [comment]
Closing as the workaround is to use the community alternative here: https://github.com/dotnet/fsharp/issues/12640#issuecomment-1028774501
